### DISCUSSION TITLE
Docker issue

### DISCUSSION THREAD

**Discussion by zalusky** [mono/SkiaSharp#2956] [Q&A]
Hello, i have a running Avalonia App which uses SkiaSharp. The VS Application is working but the Docker image does not work. In  the following i will add more data, maybe someone can help me how to modify the Dockerfile to make the project running (on x86 linux)
The project uses the CheryIUI Library which needs the follwing verions: 

```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>WinExe</OutputType>
    <TargetFramework>net8.0</TargetFramework>
    <Nullable>enable</Nullable>
    <BuiltInComInteropSupport>true</BuiltInComInteropSupport>
    <ApplicationManifest>app.manifest</ApplicationManifest>
    <AvaloniaUseCompiledBindingsByDefault>true</AvaloniaUseCompiledBindingsByDefault>
  </PropertyGroup>

  <ItemGroup>
    <Folder Include="Models\" />
    <AvaloniaResource Include="Assets\**" />
  </ItemGroup>

    <ItemGroup>
      <PackageReference Include="Avalonia.Skia" Version="11.1.0-beta2" />
      <PackageReference Include="Avalonia.Themes.Simple" Version="11.1.0-beta2" />
      <PackageReference Include="CherylUI" Version="1.2.6" />
      <PackageReference Include="ReactiveUI" Version="19.6.1" />
      <PackageReference Include="SkiaSharp" Version="3.0.0-preview.3.1" />
    </ItemGroup>
  <ItemGroup>
    <PackageReference Include="Avalonia" Version="11.1.0-beta2" />
    <PackageReference Include="Avalonia.Desktop" Version="11.1.0-beta2" />
    <PackageReference Include="Avalonia.Themes.Fluent" Version="11.1.0-beta2" />
    <PackageReference Include="Avalonia.Fonts.Inter" Version="11.1.0-beta2" />
    <!--Condition below is needed to remove Avalonia.Diagnostics package from build output in Release configuration.-->
    <PackageReference Condition="'$(Configuration)' == 'Debug'" Include="Avalonia.Diagnostics" Version="11.1.0-beta2" />
    <PackageReference Include="Avalonia.ReactiveUI" Version="11.1.0-beta2" />
  </ItemGroup>
</Project>
```

This is the dockerfile ( i have already tried playing around with different libraries...):
```Dockerfile
FROM mcr.microsoft.com/dotnet/runtime:8.0 AS base
WORKDIR /app
COPY --from=build-env /app/out .


# Install necessary dependencies for Avalonia GUI applications
RUN apt-get update && apt-get install -y --allow-unauthenticated \
#    libx11-dev \
#    libxrandr-dev \
     libc6-dev \
     libgdiplus \
     libx11-dev \
     libfreetype6 \
#    libxinerama-dev \
#    libxcursor-dev \
#    libxrender-dev \
#    libxext-dev \
#    libxi-dev \
     libfontconfig1 \
     libjpeg62-turbo \
     libpng16-16 \
     zlib1g \
     libx11-6 \
     libxext6 \
     libxrandr2 \
     libxinerama1 \
     libxcursor1 \
     libxi6 \
#    libcdev \
#    zlib1g \
#    libskiasharp \
#    libfontconfig \
    libsm6 \
    libice6

# Expose the port the app runs on
EXPOSE 5000

# Define the entry point for the container
ENTRYPOINT ["dotnet", "CheryIUI_01.dll"]
```

And this is the error message:

```
root@osboxes:/home/osboxes/Projekt/CheryIUI_01/CheryIUI_01# docker run -it --rm -e DISPLAY=$DISPLAY -v /tmp/.X11-unix:/tmp/.X11-unix cheryiui
Unhandled exception. System.TypeInitializationException: The type initializer for 'SkiaSharp.SKFontManager' threw an exception.
 ---> System.TypeInitializationException: The type initializer for 'SkiaSharp.SKObject' threw an exception.
 ---> System.InvalidOperationException: The version of the native libSkiaSharp library (88.1) is incompatible with this version of SkiaSharp. Supported versions of the native libSkiaSharp library are in the range [115.0, 116.0).
   at SkiaSharp.SkiaSharpVersion.CheckNativeLibraryCompatible(Version minSupported, Version current, Boolean throwIfIncompatible)
   at SkiaSharp.SkiaSharpVersion.CheckNativeLibraryCompatible(Boolean throwIfIncompatible)
   at SkiaSharp.SKObject..cctor()
   --- End of inner exception stack trace ---
   at SkiaSharp.SKObject..ctor(IntPtr handle, Boolean owns)
   at SkiaSharp.SKFontManager..ctor(IntPtr handle, Boolean owns)
   at SkiaSharp.SKFontManager.SKFontManagerStatic..ctor(IntPtr x)
   at SkiaSharp.SKFontManager..cctor()
   --- End of inner exception stack trace ---
   at SkiaSharp.SKFontManager.get_Default()
   at Avalonia.Skia.FontManagerImpl..ctor()
   at Avalonia.Skia.SkiaPlatform.Initialize(SkiaOptions options)
   at Avalonia.SkiaApplicationExtensions.<>c.<UseSkia>b__0_0()
   at Avalonia.AppBuilder.SetupUnsafe()
   at Avalonia.AppBuilder.Setup()
   at Avalonia.AppBuilder.SetupWithLifetime(IApplicationLifetime lifetime)
   at Avalonia.ClassicDesktopStyleApplicationLifetimeExtensions.StartWithClassicDesktopLifetime(AppBuilder builder, String[] args, Action`1 lifetimeBuilder)
   at CheryIUI_01.Program.Main(String[] args) in /app/Program.cs:line 13
```

Thank you very much in advance for any help!

**Comment by mattleibow** [comment]
The real error message is:

>  System.InvalidOperationException: The version of the native libSkiaSharp library (88.1) is incompatible with this version of SkiaSharp. Supported versions of the native libSkiaSharp library are in the range [115.0, 116.0).
 
I see you have installed SkiaSharp 3.0, but that is just the managed binary. The native bits are in [SkiaSharp.NativeAssets.Linux](https://www.nuget.org/packages/SkiaSharp.NativeAssets.Linux/)

You will have to just install that package too.

**Reply by zalusky** [reply]
Hello and thank you for your message. I have included the mentioned package, unfortunately i still receive the same error message, independent of  using a docker image or a cross compiled project for linux. Here is the modified project file:

<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>WinExe</OutputType>
    <TargetFramework>net8.0</TargetFramework>
    <Nullable>enable</Nullable>
    <BuiltInComInteropSupport>true</BuiltInComInteropSupport>
    <ApplicationManifest>app.manifest</ApplicationManifest>
    <AvaloniaUseCompiledBindingsByDefault>true</AvaloniaUseCompiledBindingsByDefault>
  </PropertyGroup>

  <ItemGroup>
    <Folder Include="Models\" />
    <AvaloniaResource Include="Assets\**" />
  </ItemGroup>

    <ItemGroup>
      <PackageReference Include="Avalonia.Skia" Version="11.1.0-beta2" />
      <PackageReference Include="Avalonia.Themes.Simple" Version="11.1.0-beta2" />
      <PackageReference Include="CherylUI" Version="1.2.6" />
      <PackageReference Include="ReactiveUI" Version="19.6.1" />
      <PackageReference Include="SkiaSharp" Version="3.0.0-preview.3.1" />
    </ItemGroup>
  <ItemGroup>
    <PackageReference Include="Avalonia" Version="11.1.0-beta2" />
    <PackageReference Include="Avalonia.Desktop" Version="11.1.0-beta2" />
    <PackageReference Include="Avalonia.Themes.Fluent" Version="11.1.0-beta2" />
    <PackageReference Include="Avalonia.Fonts.Inter" Version="11.1.0-beta2" />
    
    <PackageReference Condition="'$(Configuration)' == 'Debug'" Include="Avalonia.Diagnostics" Version="11.1.0-beta2" />
    <PackageReference Include="Avalonia.ReactiveUI" Version="11.1.0-beta2" />
    <PackageReference Include="SkiaSharp.NativeAssets.Linux" Version="2.88.8" />
  </ItemGroup>
</Project>

Maybe there are other hints which can help to solve this issue?

Thank you in advance
Gerhard
